src/controllers/estimation/estimation.ts

```diff
@@ -0,0 +1,267 @@
+    // In the case of operating with a smart account (an account with creation code), all other EOAs can pay the fee.
+    //
+    // If the current account is an EOA, only this account can pay the fee,
+    // and there's no need for checking other EOA accounts native balances.
+    // This is already handled and estimated as a fee option in the estimate library, which is why we pass an empty array here.
+    //
+    // we're excluding the view only accounts from the natives to check
+    // in all cases EXCEPT the case where we're making an estimation for
+    // the view only account itself. In all other, view only accounts options
+    // should not be present as the user cannot pay the fee with them (no key)
+    const nativeToCheck = account.creation
+      ? this.#accounts.accounts
+          .filter(
+            (acc) =>
+              !acc.creation &&
+              (acc.addr === op.accountAddr ||
+                !getIsViewOnly(this.#keystore.keys, acc.associatedKeys))
+          )
+          .map((acc) => acc.addr)
+      : []
+
+    // configure the bundler switcher for the network if any
+    const bundlerSwitcher = new BundlerSwitcher(
+      network,
+      this.#getOutsideControllerStatus,
+      this.#outsideControllerNoUpdateStatuses
+    )
+    const estimation = await getEstimation(
+      baseAcc,
+      accountState,
+      op,
+      network,
+      this.#providers.providers[op.chainId.toString()],
+      feeTokens,
+      nativeToCheck,
+      bundlerSwitcher,
+      (e: ErrorRef) => {
+        if (!this) return
+        this.estimationRetryError = e
+        this.emitUpdate()
+      }
+    ).catch((e) => e)
+
+    const isSuccess = !(estimation instanceof Error)
+    if (isSuccess) {
+      this.estimation = getEstimationSummary(estimation)
+      this.error = null
+      this.status = EstimationStatus.Success
+      this.estimationRetryError = null
+      this.availableFeeOptions = this.#getAvailableFeeOptions(baseAcc)
+    } else {
+      this.estimation = null
+      this.error = estimation
+      this.status = EstimationStatus.Error
+      this.availableFeeOptions = []
+    }
+
+    // estimation.flags.hasNonceDiscrepancy is a signal from the estimation
+    // that the account state is not the latest and needs to be updated
+    if (
+      this.estimation &&
+      (this.estimation.flags.hasNonceDiscrepancy || this.estimation.flags.has4337NonceDiscrepancy)
+    )
+      // silenly continuing on error here as the flags are more like app helpers
+      this.#accounts.updateAccountState(op.accountAddr, 'pending', [op.chainId]).catch((e) => e)
+
+    this.hasEstimated = true
+    this.emitUpdate()
+  }
+
+  // it's initialized if it has estimated at least once
+  isInitialized() {
+    return this.hasEstimated
+  }
+
+  // has it estimated at least once without a failure
+  isLoadingOrFailed() {
+    return !this.hasEstimated || !!this.error
+  }
```

Comment:
Calling it `isLoadingOrFailed` is a bit confusing because it returns false when the estimation is loading (after the first load)

Reply:
I changed this logic and I will test it again